File: backend-deno/src/validator/lobbyActions/CreateGameAction.ts
import 'reflect-metadata';
import { IsEnum } from 'class-validator';
import { Expose } from 'class-transformer';

import { LobbyAction } from './LobbyAction.ts';
import { createDTOValidator } from '../../utils/createDTOValidator.ts';

const minutesSteps = [
  ...Array.from({ length: 7 }, (_, i) => 0.25 * (i + 1)),
  ...Array.from({ length: 19 }, (_, i) => i + 2),
  ...Array.from({ length: 5 }, (_, i) => 25 + i * 5),
  ...Array.from({ length: 9 }, (_, i) => 60 + i * 15),
];
const incrementSteps = [
  ...Array.from({ length: 21 }, (_, i) => i),
  ...Array.from({ length: 5 }, (_, i) => 25 + i * 5),
  ...Array.from({ length: 5 }, (_, i) => 60 + i * 30),
];

export class CreateGameAction {
  @IsEnum(['createGame'])
  @Expose()
  type!: 'createGame';

  @IsEnum(minutesSteps)
  @Expose()
  minutes!: number;

  @IsEnum(incrementSteps)
  @Expose()
  increment!: number;

  static validate = createDTOValidator(this, {
    parseStringToJSON: true,
    before: async (val, options) => {
      return await LobbyAction.validate(val, options);
    },
  });
}

File: backend/validators/create_game_action.py
from marshmallow import Schema, fields


class CreateGameActionSchema(Schema):
    type = fields.Str(required=True, validate=lambda s: s == 'createGame')
    minutes = fields.Int(required=True, validate=lambda n: n in minutes_steps)
    increment = fields.Int(
        required=True, validate=lambda n: n in increment_steps)
    private = fields.Bool(required=True)


minutes_steps = [
    *[0.25 * (i + 1) for i in range(7)],
    *[i + 2 for i in range(19)],
    *[25 + i * 5 for i in range(5)],
    *[60 + i * 15 for i in range(9)],
]
increment_steps = [
    *[i for i in range(21)],
    *[25 + i * 5 for i in range(5)],
    *[60 + i * 30 for i in range(5)],
]
